### PR TITLE
[Neuron][Kernel] NKI-based flash-attention kernel with paged KV cache

### LINKED ISSUE #11152: [RFC][Exploratory]: vLLM Neuron Backend with V1 Architecture
### Motivation.

To leverage vLLM V1 architecture change, we are trying to propose a new integration apporach for the neuron backend that seamlessly integrate with vLLM, while maintaining high-performance and taking prefix-caching as first-class feature.

### Background

(ref: https://github.com/vllm-project/vllm/issues/8779)
vLLM is on a path toward 1/ full support for torch.compile, 2/ turn on chunked prefill, prefix caching, speculative decoding by default, 3/ support more than 60 model variants. While, current neuron backend is supported via the transformers-neuronx library, which has limited support to the combination of these feature.

To support a wide range of model variants, vLLM has been maintaining the modular design with vllm.model_executor.layers module. This enables new model developers easily contribute to vLLM to support new model variants. For instance, Mistral team released pixtral-large model weights and brought pixtral-large model support with [Pixtral (](https://github.com/patrickvonplaten/vllm/commit/d394787e5268903a705850413e494ebf2ddcefb5)[vllm-project#8377](https://github.com/vllm-project/vllm/pull/8377)[)](https://github.com/patrickvonplaten/vllm/commit/d394787e5268903a705850413e494ebf2ddcefb5).

### Embracing torch.compile support

As part of Neuron SDK 2.21 release, we are able to support torch.compile with openxla backend. For instance, we can implement copy_blocks with
```python
class NeuronAttentionBackend:
    @torch.compile(backend="openxla")
    @staticmethod
    def copy_blocks(
        kv_caches: List[Tuple[torch.Tensor, torch.Tensor]],
        src_to_dists: Tuple[torch.Tensor, torch.Tensor],
    ) -> None:
        src_indices, dst_indices = src_to_dists
        for k_cache, v_cache in kv_caches:
            torch.ops.xla.dynamo_set_buffer_donor_(k_cache, True)
            k_cache[:, dst_indices] = k_cache[:, src_indices]
            torch.ops.xla.dynamo_set_buffer_donor_(v_cache, True)
            v_cache[:, dst_indices] = v_cache[:, src_indices]

if use_torch_compile:
    compiled_code = torch.compile(PallasAttentionBackend.copy_blocks, backend='openxla')
    compiled_code(kv_caches, src_to_dists)
else:
    PallasAttentionBackend.copy_blocks(kv_caches, src_to_dists)
```

### Build neuron attention backend with NKI

We may build NKI-based flash-attention with paged KV cache, as part of vllm.attention.ops module. This is similar to triton-lang based flash-attention in vLLM (ref: [triton_flash_attention.py](https://github.com/vllm-project/vllm/blob/main/vllm/attention/ops/triton_flash_attention.py)).

### Introduce forward_neuron into vllm.model_executor.layers module

Like many other backends in vLLM, the native kernel may not be highly performant on a specific hardware backend. vLLM has been building and maintaining the default behavior with forward_native function call, while hardware-specific optimization can be enabled with forward_xxx function call.

We can reuse the performant components in neuronx_distributed package to further improvement performance.
```python
@CustomOp.register("rms_norm")
class RMSNorm(CustomOp):
    """Root mean square normalization.

    Computes x -> w * x / sqrt(E[x^2] + eps) where w is the learned weight.
    """
    def forward_neuron(
        self,
        x: torch.Tensor,
        residual: Optional[torch.Tensor] = None,
    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        from neuronx_distributed.ops import NeuronFusedRMSNorm
        if NeuronFusedRMSNorm is None:
            return self.forward_native(x, residual)
        if residual is not None:
            orig_shape = x.shape
            residual += x.view(residual.shape)
            x = NeuronFusedRMSNorm.apply(residual, self.weight, self.variance_epsilon)
            return x.view(orig_shape), residual

        x = NeuronFusedRMSNorm.apply(x, self.weight, self.variance_epsilon)
        return x
```

### Development Progress

**NKI-based Flash-attention kernel**

* [x] #11277
* [x] #13245
* [ ] #13249

**End-to-end full model support**
* [x] #13246
* [x] #14085
* [x] #14391
* [ ] https://github.com/vllm-project/vllm/pull/14648

**Test infra improvement**
* [x] Test multiple scripts by running bash - #11593 #11779

### PR DESCRIPTION
**Pull request by liangfu**
### Summary

FIX #11152 

This PR introduce a NKI-based kernel that brings the support for chunked-prefill with flash-attention.

Co-authored-by: Jiangfei Duan <jfduan@outlook.com>

### CODE REVIEW

File: vllm/attention/ops/nki_flash_attn.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+    num_blocks_per_large_tile = LARGE_TILE_SZ // block_size
+
+    block_tables_sbuf = nl.full((par_dim(B_P_SIZE), num_large_k_tile),
+                                0,
+                                dtype=np.int32,
+                                buffer=nl.sbuf)
+    for j in nl.affine_range(num_large_k_tile):
+        i_p = nl.arange(num_blocks_per_large_tile)[:, None]
+        block_tables_sbuf[i_p, j] = nl.load(
+            block_tables[j * num_blocks_per_large_tile + i_p], dtype=np.int32)
```
Comment (JF-D) [inline vllm/attention/ops/nki_flash_attn.py:430]: We should ensure that the number of blocks in each large tile doesn't exceed the partition size, otherwise we'll get out-of-bounds memory accesses. We should have another check for `num_blocks_per_large_tile`: `assert num_blocks_per_large_tile <= B_P_SIZE`.

Comment (liangfu) [reply vllm/attention/ops/nki_flash_attn.py:426]: added the assertion

**Comment by simon-mo** [comment]
@robertgshaw2-neuralmagic @WoosukKwon PTAL

File: vllm/attention/ops/nki_flash_attn.py [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+                    hbm_qk_res[batch_id, head_id * q_h_per_k_h + i_q_h, :, :],
+                    qk_res_buffer[batch_id, i_q_h, :, :],
+                )
+
+    if return_debug_tensors:
+        return o, hbm_m_buffer, hbm_l_buffer, hbm_qk_res
+    return o
+
+
+def flash_attn_varlen_nkifunc(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/attention/ops/nki_flash_attn.py:630]: Do you think we could align this to the API of the unified flash attention funciton for V1?

Comment (robertgshaw2-neuralmagic) [reply vllm/attention/ops/nki_flash_attn.py:630]: e.g. in v1/attention/backends/flash_attention

```
flash_attn_varlen_func(
                q=query[:num_actual_tokens],
                k=key_cache,
                v=value_cache,
                out=output[:num_actual_tokens],
                cu_seqlens_q=attn_metadata.query_start_loc,
                max_seqlen_q=attn_metadata.max_query_len,
                cu_seqlens_k=attn_metadata.seq_start_loc,
                max_seqlen_k=attn_metadata.max_seq_len,
                softmax_scale=self.scale,
                causal=True,
                alibi_slopes=self.alibi_slopes,
                window_size=self.sliding_window,
                block_table=attn_metadata.block_table,
                softcap=self.logits_soft_cap,
            )
```

Comment (robertgshaw2-neuralmagic) [reply vllm/attention/ops/nki_flash_attn.py:630]: Aligning to this interface will reduce special cases.

**Comment by robertgshaw2-neuralmagic** [comment]
Nice work!

File: vllm/attention/ops/nki_flash_attn.py (hunk 2 above)
Comment (liangfu) [reply vllm/attention/ops/nki_flash_attn.py:630]: Great question!

There are three challenges that are blocking the alignment of the interface:
1. calling `reshape_and_cache_flash` before `flash_attn_varlen_func` can be inefficient read-after-write, due to synchronous process on a heterogeneous architecture. On neuron stack, it's better to merge cached tokens with new tokens in an asynchronous fashion. Therefore, we pass both cached KV and active KV to the flash-attention function call, so that writing to HBM won't block compuation.
2. Since slicing `num_actual_tokens` would not reduce computation or bandwidth utilization, I think it's better to slice the logits in the last attention layer.
3. We find it more efficient to compuate attention mask before each of the layer. Therefore, instead of passing sequence lengths and query lengths, we pass the pre-computed attention mask directly to the flash-attention kernel.

I'm actively looking for ideas that can help close the gap between the interfaces, without degrading performance of the kernel.

Comment (liangfu) [reply vllm/attention/ops/nki_flash_attn.py:630]: @robertgshaw2-redhat do you think the interface misalignment can be a blocking issue for merging?

**Comment by liangfu** [comment]
@simon-mo @WoosukKwon do you have any other concerns / comments ?

**Review by simon-mo — APPROVED** [review]
This is very isolated change. I'm okay with merging this

File: vllm/attention/ops/nki_flash_attn.py [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+    # TODO : support logit_bias with continuous_batching_mask
+    assert not use_causal_mask, "causal mask is not supported."
+    assert (continuous_batching_mask
+            is not None), "continuous_batching_mask input is required."
+    if continuous_batching_mask is not None:
+        assert (logit_bias_tile is
+                None), "continuous_batching_mask does not support logit_bias!"
+
+    # mask are used to only apply computation to the lower half of the matrix,
+    # which reduce the arthimetic intensity by half
```
Comment (gaocegege) [inline vllm/attention/ops/nki_flash_attn.py:113]: Here is a typo `arthimetic` breaks the CI.

**Comment by mgoin** [comment]
PR to fix the failing precommit https://github.com/vllm-project/vllm/pull/12497
